Show only directories when using --choosedir

randynobx opened the pull request:
#### ISSUE TYPE
- Improvement/feature implementation

#### CHECKLIST
- [X] The `CONTRIBUTING` document has been read **[REQUIRED]**
- [X] All changes follow the code style **[REQUIRED]**
- [X] All new and existing tests pass **[REQUIRED]**
- [ ] Changes require config files to be updated
    - [ ] Config files have been updated
- [ ] Changes require documentation to be updated
    - [ ] Documentation has been updated
- [ ] Changes require tests to be updated
    - [ ] Tests have been updated

#### DESCRIPTION
Added global_inode_filter_type setting for global filtering of file types. Uses same logic as the filter_inode_type command.

Sets the `global_inode_filter_type` to `'d'` (for directory) when --choosedir flag is used

#### MOTIVATION AND CONTEXT
Addresses request for global filtering of files when using --choosedir= in issue #821 

#### TESTING
Have done basic testing by hand to ensure functionality

Vifon commented on ranger/container/directory.py:
```diff
@@ -252,6 +252,19 @@ def hidden_filter_func(fobj):
+            def hidden_inode_filter_func(obj):
+                g_filter = self.settings.global_inode_type_filter
+                if (
+                        ('d' in g_filter and obj.is_directory) or
```
Could we either use use a few if-else branches with one `return` in each or return a single expression? Right now it's a strange hybrid.

```python
if foo:
    return True
elif bar:
    return True
else:
    return False
```
or
```python
return foo or bar
```

Not:
```python
if foo or bar:
    return True
else:   # or even no else
    return False
```

I'd slightly prefer the first version.

Vifon replied:
Also, it's almost the same function as the one used in `commands.py`, in `filter_inode_type.execute`. Maybe we can reuse it and not duplicate this code? Especially since that one uses the named constants instead of the magic strings (`self.FILTER_DIRS` vs `'d'`).

randynobx replied:
The first version would be my choice as well.

As for the code, my understanding is that the filter does not 'exist' in ranger's code until the command is run, which is a problem for this use case.

Vifon replied:
Do you mean that it's a lambda function? Let's make it a regular function and use it in both these places. This file looks like a good place for it I'd like to hear @hut's opinion.

randynobx replied:
That sounds like a good idea. I can easily refactor both the local and global inode_filters into one block of code; which would actually take care of resolving any conflicts between the two.

I'll work on that code in case. But I too would like to hear @hut's opinion.

Vifon commented on ranger/core/main.py:
```diff
@@ -119,6 +119,9 @@ def main(
         load_settings(fm, args.clean)
 
+        if args.choosedir:
+            SettingsAware.settings.global_inode_type_filter = 'd'
```
Please use a named constant instead of `'d'`.

randynobx replied:
Where should these constants be declared/stored? The value being assigned will be the same, but I can understand wanted a constant for readability and maintenance purposes.

Vifon replied:
Either `ranger.container.directory` or some new module like `ranger.core.constants`. Again, I'd like to hear @hut's opinion on this.

randynobx replied:
AFAIK (and can find via google) the accepted way to name constants in python is to use all caps (such as `FILTER_DIR`). But I am getting pylint errors complaining that this is an 'invalid variable name'?

randynobx commented:
Refactored the inode_filter_func code and moved the logic for `:filter_inode_type` from commands.py to refilter() to reduce redundant code.

Still need to decide on where to store the constants: `FILTER_DIRS`, `FILTER_FILES`, and `FILTER_LINKS`

Vifon commented:
Looks good to me. @hut, please take a look when you'll have some time.

hut commented:
Looks good, thank you!

hut commented:
somehow the setting global_inode_type_filter is set to "d" by default even when not using --chosedir. I worked around this by unsetting it explicitly in rc.conf, but this is weird anyway. Got any idea why?

randynobx commented:
I'm unable to reproduce this behavior.

Just to clarify, the filter is enabled by default (and set to "d") even without the --choosedir flag?

hut commented:
Yes. Do you use the current master branch with the workaround, maybe? Try commit 88e0bc0540fb963b3cf638217dd578b127fd6a3a

randynobx commented:
I actually haven't applied the workaround patch. I'm using the master branch at commit 88e0bc0 right now

hut commented:
Ok, I will investigate later
